### PR TITLE
Add pulp_integration framework

### PR DESCRIPTION
**Pull request by Shrews**
This adds a new `pulp_integration` test framework that will allow us to have a local image repository that supports signed images. The PR that adds support for container image signature validation (PR #371) will be able to use this to add integration tests.

Note that the Pulp Integration job fails in this PR because it is adding a new GHA job that uses a secret. It will not have access to this secret until this is merged, but I verified that it does work as a PR against my own fork, seen here: ~~https://github.com/Shrews/ansible-builder/runs/7134536766?check_suite_focus=true~~
~~https://github.com/Shrews/ansible-builder/runs/7201566704?check_suite_focus=true~~
https://github.com/Shrews/ansible-builder/runs/7254569323?check_suite_focus=true

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (nitzmahone) [inline .github/workflows/ci.yml]: This should be `python-version`- note the warning at https://github.com/Shrews/ansible-builder/runs/7134536766?check_suite_focus=true#step:3:1 . It's currently doing the right thing, but only because the default `python-version: 3.x` the action declares currently resolves to `3.10`.

File: setup_pulp.sh
Comment (nitzmahone) [inline setup_pulp.sh]: Any reason to repeat this outside the loop instead of just initing to empty (or not at all, which IIRC it won't care about anyway)?

File: setup_pulp.sh
Comment (nitzmahone) [inline setup_pulp.sh]: Given core's past experience with the pulp container DB init just going to lunch and never coming back once in awhile, might be good to add an error message if this exit trips.

File: setup_pulp.sh
Comment (nitzmahone) [inline setup_pulp.sh]: This file should probably be under .github, not at the root of the repo- maybe `.github/test-scripts` or something?

File: test/pulp_integration/test_policies.py
Comment (nitzmahone) [inline test/pulp_integration/test_policies.py:3]: I assume this is basically `# FIXME: implement`, or ?

**Review by nitzmahone — COMMENTED** [review]
A couple suggestions, but I don't see anything terribly off-the-rails here for your first GHA experience ;)

File: setup_pulp.sh
Comment (eqrx) [inline setup_pulp.sh]: Question from me: Is is still common to have the interpreter under `/bin`? Have my bash under `/usr/bin` since I remember :) (`/bin` links to `/usr/bin`)

File: setup_pulp.sh
Comment (eqrx) [inline setup_pulp.sh]: Would recommend setting `-u` to catch unset variables.

Cherry on top: `set -eu -o pipefail`

But since this script has neither vars nor I would find it understandable if we keep it like that

**Review by eqrx — CHANGES_REQUESTED** [review]

File: setup_pulp.sh
Comment (Shrews) [reply setup_pulp.sh]: I dunno. Less typing this way? I can modify it.

File: .github/workflows/ci.yml
Comment (Shrews) [reply .github/workflows/ci.yml]: Oopsie. Done

File: setup_pulp.sh
Comment (Shrews) [reply setup_pulp.sh]: Done

File: setup_pulp.sh
Comment (Shrews) [reply setup_pulp.sh]: Done

File: setup_pulp.sh
Comment (Shrews) [reply setup_pulp.sh]: Done

File: test/pulp_integration/test_policies.py
Comment (Shrews) [reply test/pulp_integration/test_policies.py:3]: Yup. This is where the next PR would pick up work, modifying this test file. Added a comment for clarity.

File: setup_pulp.sh
Comment (Shrews) [reply setup_pulp.sh]: Added `-u`

**Comment by Shrews** [comment]
Lastest working run against my fork: https://github.com/Shrews/ansible-builder/runs/7201566704?check_suite_focus=true

**Review by eqrx — APPROVED** [review]

File: .github/test-scripts/setup_pulp.sh
Comment (eqrx) [inline .github/test-scripts/setup_pulp.sh]: Maybe guard against `$1` being empty? Would require an if block, decreases readability of the code but would report something more obvious than ENOTTY when the password secret is not set. Not sure if I would actually implement that since this is a shell script and bound to fail anyway :D

Comment (Shrews) [reply .github/test-scripts/setup_pulp.sh]: Added a check to exit early if pwd is empty (should never happen once merged, really).

File: .github/test-scripts/setup_pulp.sh
Comment (nitzmahone) [inline .github/test-scripts/setup_pulp.sh]: Thinking about this one a little more- the username should probably also be an arg provided by a GHA secret- not so much that it's a secret, but if someone wanted to run this from a fork, they'd have to provide their own token, which would have a different username. Since GHA doesn't really have a general concept of external "vars" that aren't secrets, a secret it'd have to be. :(

Comment (Shrews) [reply .github/test-scripts/setup_pulp.sh]: Done.

Comment (Shrews) [reply .github/test-scripts/setup_pulp.sh]: Seen as secrets here: https://github.com/Shrews/ansible-builder/runs/7254569323?check_suite_focus=true#step:6:149

**Review by nitzmahone — APPROVED** [review]

**Comment by nitzmahone** [comment]
Merging, should go green once it has access to local secrets. We may need to retarget the pulp job to only run on maintainer-sourced PRs or merge jobs only so it doesn't fail on all future PRs.

**Pull request merged**
